man/man1/pmwtf.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2013-2014 Red Hat.
.\" 
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\" 
.\" 
.TH PMWTF 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmwtf\f1 \- compares archives and report significant differences
.SH SYNOPSIS
\f3pmwtf\f1
[\f3\-d\f1/\f3--keep\f1]
[\f3\-z\f1/\f3--hostzone\f1]
[\f3\-p\f1/\f3--precision\f1 \f2precision\f1]
[\f3\-q\f1/\f3--threshold\f1 \f2thres\f1]
[\f3\-S\f1/\f3--start\f1 \f2starttime\f1]
[\f3\-T\f1/\f3--finish\f1 \f2endtime\f1]
[\f3\-B\f1/\f3--begin\f1 \f2starttime\f1]
[\f3\-E\f1/\f3--end\f1 \f2endtime\f1]
[\f3\-x\f1 \f2metric\f1]
[\f3\-X\f1 \f2file\f1]
[\f3--skip-excluded\f1]
[\f3--skip-missing\f1]
[\f3\-Z\f1/\f3--timezone\f1 \f2timezone\f1]
\f2archive1\f1
[\f2archive2\f1]
.SH DESCRIPTION
.B pmwtf
compares the average values for every metric in either one
or two archives, in a given time window, for changes that are
likely to be of interest when searching for performance regressions.
.PP
The archive log has the base name
.I archive
and must have been previously created using
.BR pmlogger (1).
The
.BR pmlogsummary (1)
utility is used to obtain the average values used for comparison.
.PP
There are two sorts of invocation of the tool: with either one or
two archives.
.PP
In the first case, the only sensible command line requires use of
all four time window arguments.  These are specified using the same
time window format described in
.BR PCPIntro (1),
and are
.BR \-S / \-\-start
and
.BR \-T / \-\-finish
for the start and end times of the first time window of interest
in the archive, and
.BR \-B / \-\-before
and
.BR \-E / \-\-end
for the start and end times of the second time window of interest.
.PP
In the second case, with two archives, the 
.BR \-B / \-\-before
and
.BR \-E / \-\-end
options might be unnecessary.  This might be the case, for example,
when comparing the same time window of two consecutive days (usually
two separate archives), or a time window on the same day of different
weeks.
.PP
In either case,
.B pmwtf
produces a sorted summary of those metrics in the specified window
whose values have deviated the most from a minimal threshold.
The level of deviation is calculated by dividing the average value
of each metric in both logs, and then calculating whether the ratio
falls outside of a range considered normal.
This ratio can be adjusted using the
.BR \-q / \-\-threshold
option, and by default it is 2 (i.e. report all metrics with average
values that have more than doubled in the two time windows or more
than halved in the two time windows).
.PP
Should any metrics be present in one window but missing from the
other, a diagnostic will be displayed listing each missing metric
and the archive from which it was missing.
.PP
The remaining options control the specific information to be reported.
Metrics with counter semantics are converted to rates before being
evaluated.
.TP 5
.BR \-p / \-\-precision
Print all floating point numbers with 
.I precision
digits after the decimal place.
.TP
.B \-\-skip-excluded
Cull the list of names of metrics being excluded from the output.
.TP
.B \-\-skip-missing
By default,
.B pmwtf
will report the names of any metrics that are in one archive but not
the other.
This option suppresses that reporting.
.TP
.B \-x
Compare each metric in each archive in the time windows specified
to a given
.BR egrep (1)
pattern, excluding those that match from the report output.
.TP
.B \-X
Allows a
.IR file
to be specified which containing
.BR egrep (1)
patterns which are applied to the metric names to optionally exclude
some from the report.
.TP
.B \-z
Use the local timezone from the given archives.
.TP
.BR \-Z / \-\-timezone
Changes the timezone in the archive labels to
.I timezone
in the format of the environment variable
.B TZ
as described in
.BR environ (5).
.PP
.SH FILES
.PD 0
.TP 10
.BI $PCP_LOG_DIR/pmlogger/ hostname
Default directory for PCP archives containing performance
metric values collected from the host
.IR hostname .
.PD
.SH "PCP ENVIRONMENT"
Environment variables with the prefix
.B PCP_
are used to parameterize the file and directory names
used by PCP.
On each installation, the file
.I /etc/pcp.conf
contains the local values for these variables.
The
.B $PCP_CONF
variable may be used to specify an alternative
configuration file,
as described in
.BR pcp.conf (5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pmlogger (1),
.BR pmlogsummary (1),
.BR egrep (1),
.BR pcp.conf (5)
and
.BR pcp.env (5).


CHANGELOG
pcp-3.9.10 (5 September 2014)
    - Currently under development.

pcp-3.9.9 (13 August 2014)
    - sar2pcp: support for systat version 10.2.0
    - pmiostat(1): new monitor tool with iostat-alike output
    - pcp-dmcache(1): new monitor tool for dm-cache devices
    - pmdalinux: disk.dm.* metric hierarchiy (device mapper)
    - pmdalinux: refine device mapper naming for old platforms
    - pmdalinux: allow controlled data injection for QA tests
    - pmdadmcache: new agent for device mapper cache targets
    - pmdainfiniband: correctly initialise switch metrics
    - libpcp: global timeout option for service discovery
    - libpcp: __pmDiscoverServicesWithOptions() improvements
    - libpcp: grok '%<interface' for link-local ipv6 addresses
    - libpcp: derived metrics support for new rate() keyword
    - python api: many convenience class module extensions
    - python api: timeval accessor helper interfaces
    - python api: pmUnitStr now uses thread-safe version
    - python api: pmNumberStr_r now available from python
    - python api: fix derived metrics interface exceptions
    - pmlogconf: mark some log-default metrics as log-once
    - dbpmda: use the correct (blocking) socket API variants
    - pmlogger_daily: fix $PCP_RUN_DIR creation checks

pcp-3.9.8 (16 July 2014)
    - Add overlooked Debian packaging metadata updates.

pcp-3.9.7 (16 July 2014)
    - pmie: "ruleset" extensions to the pmie(1) language
    - pmdaproc: additional memory and new blkio cgroup metrics
    - pmdanvidia: new Nvidia GPU PMDA using NVML interfaces
    - pmdainfiniband: add switch counter metrics
    - pmdalustrecomm: bug fixes in the lnet memory metrics
    - pmdamemcache: fixed underlying socket code
    - pmdalinux: mem.util.available for down-rev kernels
    - pmdagfs2: several fixes and improvements in behaviour
    - pmdas: support for long form command line options
      (aix, jbd2, linux, lustre, mounts, processes, sample,
      sendmail, txmon, xfs)
    - libpcp: 'timeout' option for 'probe' service discovery
    - libpcp: secure sockets timeout handling improvements
    - pmie examples: updates to the pmie tutorial docs
    - pmchart: new MemAvailable view for mem.util.available
    - porting: additional GNU/Hurd port updates
    - packaging: fix Makepkgs script for old tar versions
    - packaging: fix rpm builds for rpm versions prior to 4.6
    - libpcp: use minimum required stack for discovery threads
    - libpcp: archive folio parsing in option handling code
    - python APIs: archive folio parsing for recording tools
    - man pages: updates to service discovery APIs and tools
    - libpcp: archive interpolation thread safety improvements
    - libpcp: fixed memory leak from pmDestroyContext(3)

pcp-3.9.6 (19 June 2014)
    - Fix pthreads build fallout for Solaris platforms
    - Remove accidental /var/run/pcp install on Debian
    - Rename configure.in to configure.ac as autoconf prefers

pcp-3.9.5 (18 June 2014)
    - porting work on the GNU/Hurd platform
    - porting work on Windows (early MinGW-64 work)
    - pmdalinux: add in the new mem.util.available metric
    - pmchart: new vCPU view, resolve guest time reporting
    - pmwebd: add CORS "Access-Control-Allow-Origin: *" to
      JSON response headers
    - pmatop: improvements to value fitting
    - packaging: gui debuginfo/testsuite correctly replaced
    - pmconfig: moved install location for Solaris conflict
    - pmstat: fix hostname reporting when in cluster mode
    - Coverity scan fixes in libpcp, python, and elsewhere
    - packaging: pcp-doc now contains all of the tutorial bits
    - python: timeval/metricspec error handling improvements
    - pmatop: handle curses problems more gracefully
    - pmdapostfix: run as postfix user, not the pcp user
    - python: fix pmParseInterval failure segfault
    - python: fix segfault when printing empty units
    - libpcp: functions for collecting server port numbers
    - pmcd.services metric reflecting PCP daemon usage
    - libpcp, pmfind: active service probing
    - pmproxy, pmwebd: support for service discovery
    - pmcollectl, pmatop: improved log writing support
    - pmie: long option support
    - pmquery: neater output formatting
    - pmchart, pmtime: long option support
    - pmdasystemd: install the built binary
    - python: significant work on the pcp.pmcc module, adding
      long option support, scale conversion, an initial fetch
      loop method, and entirely new convenience classes
    - pmlogcheck: fix bad memory access sigsegv on some logs
    - Fix infinite loop in build scripts without Qt enabled
    - pmdalinux: add metrics from /proc/net/netstat
    - pmdumptext: long option support
    - pmcollectl: support for sub-second sampling intervals
    - Revive support for RHEL5-era builds, post pcp-gui merge

pcp-3.9.4 (15 May 2014)
    - pmlogger: improved archive I/O features, esp for access to
      archives while they are actively being written
    - pmlogger/pmlc: all flushing operations are now no-ops, but
      they remain available for backward compatibility
    - pcp-free: use pmConvScale in case metric units ever change
    - pmlogmv: fix a couple of argument handling corner cases
    - another round of IPv6 updates throughout PCP, making use
      of __pm* networking APIs instead of e.g. gethostbyname.
    - archive utilities updated for long option support:
      pmdumplog, pmlogcheck, pmlogextract, pmlogger, pmloglabel,
      pmlogreduce, pmlogrewrite and pmlogsummary
    - PMDAs with long option support:
      pmdaapache, pmdabash, pmdagfs2, pmdasample, pmdasimple,
      pmdatrivial
    - libpcp: allow addition of free-form text to usage messages
    - pmgetopt: new cross-platform long option handling utility,
      allowing pcp shell scripts to acquire long option support
    - shell scripts updated to support long options:
      pcp, pmie2col, pmie_check, pmie_daily, pmlogger_check,

src/pmlogsummary/pmwtf.sh
#!/bin/sh
#
# Copyright (c) 2014 Red Hat.
# Copyright (c) 2008-2010 Aconex.  All Rights Reserved.
# 
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
# 
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
# 
# Compare two PCP archives and report significant differences
#

# Get standard environment
. $PCP_DIR/etc/pcp.env

tmp=`mktemp -d /var/tmp/pcp.XXXXXXXXX` || exit 1
status=1
trap "rm -rf $tmp; exit \$status" 0 1 2 3 15
prog=`basename $0`

cat > $tmp/usage << EOF
# Usage: [options] archive1 [archive2]

Options:
  -d,--keep            debug, keep intermediate files
  -p=N,--precision=N   number of digits to display after the decimal point
  -q=N,--threshold=N   change interesting threshold to be > N or < 1/N [N=2]
  --skip-missing       do not report metrics missing between the archives
  --skip-excluded      do not report the list of metrics being excluded
  --start
  --finish
  -B=TIME,--begin=TIME start time for second archive (optional)
  -E=TIME,--end=TIME   end time for second archive (optional)
  -x=REGEX             egrep(1) pattern of metric(s) to be excluded
  -X=FILE              file containing egrep(1) patterns to exclude
  --timezone
  --hostzone
  --help
EOF

_usage()
{
    pmgetopt --usage --progname=$prog --config=$tmp/usage
    exit $status
}

_fix()
{
    sed -e 's/  *\([0-9][0-9.]*\)\([^"]*\)$/|\1/' \
    | egrep -v -f $tmp/exclude \
    | sort -t\| -k1,1
}

cat <<'End-of-File' >$tmp/exclude
^pmcd.pmlogger.port 
End-of-File

thres=2
opts=""
start1=""
start2=""
finish1=""
finish2=""
precision=3
skip_missing=false
skip_excluded=false

ARGS=`pmgetopt --progname=$prog --config=$tmp/usage -- "$@"`
[ $? != 0 ] && exit 1

eval set -- "$ARGS"
while [ $# -gt 0 ]
do
    case "$1"
    in
	-d)	trap "exit \$status" 0 1 2 3 15
		otmp="$tmp"
		tmp=`pwd`/tmp
		[ -d "$tmp" ] || mkdir "$tmp" || exit 1
		mv $otmp/exclude $tmp/exclude
		rmdir $otmp
		;;
	-p)	precision="$2"
		shift
		opts="$opts -p $precision"
		;;
	-q)	thres="$2"
		shift
		;;
	-S)	start1="$2"
		shift
		;;
	-T)	finish1="$2"
		shift
		;;
	-B)	start2="$2"
		shift
		;;
	-E)	finish2="$2"
		shift
		;;
	-x)	echo "$2" >>$tmp/exclude
		shift
		;;
	-X)	cat "$2" >>$tmp/exclude
		shift
		;;
	-z)	opts="$opts -z"
		;;
	-Z)	opts="$opts -Z $2"
		shift
		;;
	--skip-missing)
		skip_missing=true
		;;
	--skip-excluded)
		skip_excluded=true
		;;
	--)	shift
		break
		;;
	-\?)	_usage
		# NOTREACHED
		;;
    esac
    shift
done

if [ $# -lt 1 -o $# -gt 2 ]
then
    _usage
    # NOTREACHED
elif [ $# -eq 2 ]
then
    arch1="$1"
    arch2="$2"
else
    arch1="$1"
    arch2="$1"
fi

[ $precision -lt 3 -o $precision -gt 15 ] && precision=3
colwidth=`expr 12 + $precision`

echo "Directory: `pwd`"
if ! $skip_excluded
then
    echo "Excluded metrics:"
    sed -e 's/^/    /' <$tmp/exclude
fi
echo

options="$opts"
if [ "X$start1" != X ]; then
    options="$options -S $start1"
fi
if [ "X$finish1" != X ]; then
    options="$options -T $finish1"
fi
pmlogsummary -N $options $arch1 2>$tmp/err | _fix >$tmp/1
if [ -s $tmp/err ]
then
    echo "Warnings from pmlogsummary ... $arch1"
    cat $tmp/err
    echo
fi

options="$opts"
if [ "X$start2" != X ]; then
    options="$options -S $start2"
elif [ "X$start1" != X ]; then
    options="$options -S $start1"
fi
if [ "X$finish2" != X ]; then
    options="$options -T $finish2"
elif [ "X$finish1" != X ]; then
    options="$options -T $finish1"
fi
pmlogsummary -N $options $arch2 2>$tmp/err | _fix >$tmp/2
if [ -s $tmp/err ]
then
    echo "Warnings from pmlogsummary ... $arch2"
    cat $tmp/err
    echo
fi

if [ -z "$start1" ] 
then
    window1="start"
else
    window1="$start1"
fi
if [ -z "$finish1" ] 
then
    window1="$window1-end"
else
    window1="$window1-$finish1"
fi
if [ -z "$start2" ] 
then
    window2="start"
else
    window2="$start2"
fi
if [ -z "$finish2" ] 
then
    window2="$window2-end"
else
    window2="$window2-$finish2"
fi

if ! $skip_missing
then
    join -t\| -v 2 $tmp/1 $tmp/2 >$tmp/tmp
    if [ -s $tmp/tmp ]
    then
	echo "Missing from $arch1 $window1 (not compared) ..."
	sed <$tmp/tmp -e 's/|.*//' -e 's/^/    /'
	echo
    fi

    join -t\| -v 1 $tmp/1 $tmp/2 >$tmp/tmp
    if [ -s $tmp/tmp ]
    then
	echo "Missing from $arch2 $window2 (not compared) ..."
	sed <$tmp/tmp -e 's/|.*//' -e 's/^/    /'
	echo
    fi
fi

a1=`basename "$arch1"`
a2=`basename "$arch2"`
echo "$thres" | awk '
    { printf "Ratio Threshold: > %.2f or < %.3f\n",'"$thres"',1/'"$thres"'
      printf "%*s %*s   Ratio  Metric-Instance\n",
             '$colwidth', "'"$a1"'", '$colwidth', "'"$a2"'" }'
if [ -z "$start1" ] 
then
    window1="start"
else
    window1="$start1"
fi
if [ -z "$finish1" ] 
then
    window1="$window1-end"
else
    window1="$window1-$finish1"
fi
if [ -z "$start2" ] 
then
    window2="start"
else
    window2="$start2"
fi
if [ -z "$finish2" ] 
then
    window2="$window2-end"
else
    window2="$window2-$finish2"
fi
printf '%*s %*s\n' $colwidth "$window1" $colwidth "$window2"
join -t\| $tmp/1 $tmp/2 \
| awk -F\| '
function doval(v)
{
    precision='"$precision"'
    extra=precision-3
    if (v > 99999999)
	printf "%*.*f%*s",15+extra,0,v,1," "
    else if (v > 999)
	printf "%*.*f%*s",11,0,v,2+precision," "
    else if (v > 99)
	printf "%*.*f%*s",13+extra,1+extra,v,3," "
    else if (v > 9)
	printf "%*.*f%*s",14+extra,2+extra,v,2," "
    else
	printf "%*.*f%*s",15+extra,precision,v,1," "
}
$3+0 == 0 || $2+0 == 0 {
		if ($3 == $2)
		    next
		doval($2)
		doval($3)
		printf "   "
		if ($3+0 == 0)
		    printf "|-|   %s\n",$1
		else if ($2+0 == 0)
		    printf "|+|   %s\n",$1
		next
}
$2 / $3 > '"$thres"' || $3 / $2 > '"$thres"'	{
		doval($2)
		doval($3)
		printf " "
		r = $3/$2
		if (r < 0.001)
		    printf " 0.001-"
		else if (r < 0.01)
		    printf "%6.3f ",r
		else if (r > 100)
		    printf " 100+  "
		else
		    printf "%5.2f  ",r
		printf " %s\n",$1
	}' \
| sort -nr -k 3,3 \
| sed -e 's/100+/>100/' -e 's/ 0.001-/<0.001 /'

status=0
exit
